### PR TITLE
Implement real quantized tanh

### PR DESCRIPTION
**Pull request by laurilaatu**
Implement a way to use tanh from Keras backend and use `K.cast_to_floatx` in `call`. Preserve the previous functionality with integer and symmetric parts. This is useful when quantizing models that have previously been trained using the tanh from Keras.

Mentioned in issue #62 

quantized_tanh_real

### CODE REVIEW

**Review by zhuangh — CHANGES_REQUESTED** [review]
again, nice figure! but have you noticed the asymmetry around -2 and 2?

**Comment by laurilaatu** [comment]
> again, nice figure! but have you noticed the asymmetry around -2 and 2?

Thanks for the comments.

This implementation keeps the possibility for `symmetric` from the original implementation.

Here is how it looks when setting `symmetric = 1`

quantized_tanh_real_symmetric

File: qkeras/quantizers.py [hunk 1]
```diff
@@ -1643,38 +1643,48 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
...
   Returns:
     Function that performs tanh + quantization to bits in the range -1.0 to 1.0.
   """
 
   def __init__(self, bits=8, integer=0, symmetric=0,
-               use_stochastic_rounding=False):
+               use_stochastic_rounding=False, use_real_tanh=False):
     super(quantized_tanh, self).__init__()
     self.bits = bits
     self.integer = integer
```
Comment (zhuangh) [inline qkeras/quantizers.py:1659]: after discussion, we want to remove this integer. Could you help remove it?

**Review by zhuangh — CHANGES_REQUESTED** [review]
the figure looks nicer!

File: qkeras/quantizers.py (hunk 1 above)
Comment (zhuangh) [reply qkeras/quantizers.py:1659]: Hi @laurilaatu could you also add the unittests for quantized_tanh, thanks

**Comment by laurilaatu** [comment]
This version has integer removed and is also always symmetric reaching -1 and 1 similarly to `quantized_sigmoid`. One discrepancy between the two tanhs is that computing `tanh` as `2*sigmoid-1` has double the step size compared to using `K.tanh`.

Let me know if I should do further changes. Once the implementation is good I will continue to write a test for this.


tanh_no_int

File: qkeras/quantizers.py [hunk 2]
```diff
@@ -1641,57 +1641,47 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
...
-        (_round_through(p, self.use_stochastic_rounding) / m) - 1.0, -1.0 +
-        (1.0 * self.symmetric) / m, 1.0 - 1.0 / m)
-    return xq
+    x = K.cast_to_floatx(x)
+    m = K.cast_to_floatx(K.pow(2, non_sign_bits))
+    if not self.use_real_tanh:
+      xi = 2.0 * (_round_through(_sigmoid(x) * m,
+                  self.use_stochastic_rounding) / m) - 1.0
+    else:
+      xi = (_round_through(K.tanh(x) * m,
```
Comment (danielemoro) [inline qkeras/quantizers.py:1674]: The standard is to call this xq

File: qkeras/quantizers.py [hunk 3]
```diff
@@ -1641,57 +1641,47 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
...
+    if self.use_real_tanh:
+      flags.append(str(int(self.use_real_tanh)))
     return "quantized_tanh(" + ",".join(flags) + ")"
 
   def __call__(self, x):
     non_sign_bits = self.bits - 1
-    m = pow(2, non_sign_bits)
-    m_i = pow(2, self.integer)
-    p = _sigmoid(x / m_i) * m
-    xq = m_i * tf.keras.backend.clip(
```
Comment (danielemoro) [inline qkeras/quantizers.py:1674]: With  `quantized_tanh(4,use_real_tanh=True)`, I'm seeing 17 unique values, which is more than 2^4=16. This means that there must be a bug since we have more levels than can be represented with 4 bits. I think the key missing part is the clipping function

Perhaps try something like this:
```
    if not self.use_real_tanh:
      p = _sigmoid(x) * m
      xq = tf.keras.backend.clip(2.0 * (_round_through(p, self.use_stochastic_rounding) / m) - 1.0,
                                       -1.0 + 1.0 / m,
                                       1.0 - 1.0 / m)
    else:
      p = K.tanh(x) * m
      xq = tf.keras.backend.clip((_round_through(p, self.use_stochastic_rounding) / m),
                                       -1.0 + 1.0 / m,
                                       1.0 - 1.0 / m)
```

**Review by danielemoro — CHANGES_REQUESTED** [review]
This looks great!

File: qkeras/quantizers.py (hunk 3 above)
Comment (danielemoro) [reply qkeras/quantizers.py:1674]: I just realized that even with the code I linked, the real `tanh` has more steps than the `_sigmoid`. This code should fix that issue such that both `use_real_tanh = True` and `use_real_tanh = False` have the same number of steps:

```
def __call__(self, x):
    non_sign_bits = self.bits - 1
    x = K.cast_to_floatx(x)
    m = K.cast_to_floatx(K.pow(2, non_sign_bits))
    p = K.tanh(x) if self.use_real_tanh else 2.0 * _sigmoid(x) - 1.0
    xq = tf.keras.backend.clip((_round_through(p * m, self.use_stochastic_rounding) / m),
                                -1.0 + 1.0 / m,
                                1.0 - 1.0 / m)
    return xq

Comment (laurilaatu) [reply qkeras/quantizers.py:1674]: Thanks for the suggestion!

**Comment by laurilaatu** [comment]
The implementation now clips so that there is not too many values. I also fixed `min()` and `max()` to reflect this change. If this looks good I will fix the test for `qrnn`.

tanh_clip


This issue was present in `quantized_sigmoid` so I added the clip there too which now looks like this:

sigmoid_clip

File: tests/qactivation_test.py [hunk 4]
```diff
@@ -301,6 +302,45 @@ def test_quantized_sigmoid(bits, sigmoid_type, use_real_sigmoid, test_values, ex
...
+                [[-1.  , -0.75, -0.5 , -0.25,  0.  ,  0.25,  0.5 ,  0.75]],
+                dtype=K.floatx()),
+            np.array([[-0.75 , -0.625, -0.5  , -0.25 ,  0.   ,  0.25 ,  0.5  ,  0.625]],
+                     dtype=K.floatx()),
+        )
+    ])
+def test_quantized_tanh(bits, use_real_tanh, test_values, expected_values):
+  """Test quantized_sigmoid function with three different sigmoid variants."""
+  # store previous sigmoid type
+  if quantized_sigmoid(5)(K.cast_to_floatx([1.0])).numpy()[0] == 0.96875:
```
Comment (danielemoro) [inline tests/qactivation_test.py:329]: This is not a reliable way to test what the previous sigmoid type was since it is reliant on the testing input. I suggest this approach instead:

In the `quantizers.py` file, set a default sigmoid type on line 204 like so:
```
_default_sigmoid_type = "hard"
_sigmoid = None


def set_internal_sigmoid(mode):
  """Sets _sigmoid to either real, hard or smooth."""

  global _sigmoid

  if mode not in ["real", "hard", "smooth"]:
    raise ValueError("mode has to be 'hard' or 'smooth'.")

  if mode == "hard":
    _sigmoid = hard_sigmoid
  elif mode == "smooth":
    _sigmoid = smooth_sigmoid
  elif mode == "real":
    _sigmoid = tf.keras.backend.sigmoid

set_internal_sigmoid(_default_sigmoid_type)
```

Then in `qactivation_test.py`, you can simply import `_default_sigmoid_type` and set the sigmoid type to that at the end of every test like so:
```
set_internal_sigmoid(_default_sigmoid_type)
```

File: qkeras/quantizers.py [hunk 5]
```diff
@@ -1742,11 +1728,14 @@ def __call__(self, x):
       p = K.sigmoid(x) * m
     else:
       p = _sigmoid(x) * m
-    return (_round_through(p, self.use_stochastic_rounding) / m)
+
+    return tf.keras.backend.clip((_round_through(p, self.use_stochastic_rounding) / m),
+                                 0,
```
Comment (danielemoro) [inline qkeras/quantizers.py:1733]: This `0` should instead be `1.0 / m` to enforce symmetry

File: qkeras/quantizers.py [hunk 6]
```diff
@@ -1641,57 +1641,44 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
...
+                                 1.0 - 1.0 / m)
 
   def max(self):
     """Get the maximum value that quantized_tanh can represent."""
-    unsigned_bits = self.bits - 1
-    if unsigned_bits > 0:
-      return max(1.0, np.power(2.0, self.integer))
-    else:
-      return 1.0
+    return 1.0 - 1.0 / pow(2, self.bits - 1)
```
Comment (danielemoro) [inline qkeras/quantizers.py:1677]: It may be worth adding a test for the max and min methods.

File: qkeras/quantizers.py [hunk 7]
```diff
@@ -1742,11 +1728,14 @@ def __call__(self, x):
       p = K.sigmoid(x) * m
```
Comment (danielemoro) [inline qkeras/quantizers.py:1728]: To mirror the quantized_tanh, I suggest setting the `p` variable in the same way you set the `p` variable in `quantized_tanh`:
```
p = K.sigmoid(x) if self.use_real_sigmoid else _sigmoid(x)
```

This may require some other small adjustments in this call function.

File: qkeras/quantizers.py [hunk 8]
```diff
@@ -1641,57 +1641,44 @@ class quantized_tanh(BaseQuantizer):  # pylint: disable=invalid-name
...
-    p = _sigmoid(x / m_i) * m
-    xq = m_i * tf.keras.backend.clip(
-        2.0 *
-        (_round_through(p, self.use_stochastic_rounding) / m) - 1.0, -1.0 +
-        (1.0 * self.symmetric) / m, 1.0 - 1.0 / m)
-    return xq
+    x = K.cast_to_floatx(x)
+    m = K.cast_to_floatx(K.pow(2, non_sign_bits))
+    p = K.tanh(x) if self.use_real_tanh else 2.0 * _sigmoid(x) - 1.0
+    return tf.keras.backend.clip((_round_through(p * m, self.use_stochastic_rounding) / m),
```
Comment (danielemoro) [inline qkeras/quantizers.py:1671]: This line is too long

**Review by danielemoro — CHANGES_REQUESTED** [review]
This looks great! I added some comments, but the overall approach looks correct to me. 
Please also go ahead and fix the remaining failing tests. Thank you! 

I apologize for the late review.

File: tests/quantizer_impl_test.py [hunk 9]
```diff
@@ -50,7 +50,6 @@ def test_QuantizedTanh():
   qtools_quantizer = quantizer_impl.QuantizedTanh()
   qtools_quantizer.convert_qkeras_quantizer(qkeras_quantizer)
   new_quantizer = qtools_quantizer.convert_to_qkeras_quantizer(
-      symmetric=qkeras_quantizer.symmetric,
```
Comment (danielemoro) [inline tests/quantizer_impl_test.py:53]: Why was this line removed?

File: qkeras/qtools/quantized_operators/quantizer_impl.py [hunk 10]
```diff
@@ -117,16 +117,14 @@ def convert_qkeras_quantizer(
       self, quantizer: quantizers.quantized_tanh):
     self.mode = 0
     self.bits = quantizer.bits
-    self.int_bits = quantizer.integer
     self.is_signed = 1
 
   def convert_to_qkeras_quantizer(
-      self, symmetric=0, use_stochastic_rounding=False):
```
Comment (danielemoro) [inline qkeras/qtools/quantized_operators/quantizer_impl.py:124]: Why was this arg removed?

**Review by danielemoro — COMMENTED** [review]
I think it is almost there. I just have some questions

File: tests/quantizer_impl_test.py (hunk 9 above)
Comment (laurilaatu) [reply tests/quantizer_impl_test.py:53]: I had removed `symmetric` along with `integer`. I have reintroduced `symmetric` in the latest commit.

**Review by danielemoro — APPROVED** [review]

**Review by zhuangh — APPROVED** [review]
